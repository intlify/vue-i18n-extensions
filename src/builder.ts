import { isNumber } from '@intlify/shared'

/**
 * Content Builder options
 *
 * @remarks
 * options that creating a {@link ContentBuilder}
 *
 * @public
 */
export interface ContentBuilderOptions {
  /**
   * Intdent level of Builder
   */
  indentLevel?: number
}

/**
 * Content Builder
 *
 * @public
 */
export interface ContentBuilder {
  /**
   * Indent level
   * @defaultValue 0
   */
  readonly indentLevel: number
  /**
   * Content
   */
  readonly content: string
  /**
   * Add content
   * @param content - additional content
   */
  push(content: string): void
  /**
   * Add line break
   */
  newline(): void
  /**
   * Add content with line break
   * @param content - additional content
   */
  pushline(content: string): void
  /**
   * Indent content
   * @param withNewLine - whether indent to be added without new line
   */
  indent(withoutNewLine?: boolean): void
  /**
   * DeIndent content
   * @param withoutNewLine - whether deindent to be added without new line
   */
  deindent(withoutNewLine?: boolean): void
}

/**
 * Create a Content Builder
 *
 * @param options - Content Builder options
 *
 * @returns A {@link ContentBuilder} instance
 *
 * @public
 */
export function createContentBuilder(
  options: ContentBuilderOptions = {}
): ContentBuilder {
  let _indentLevel =
    options.indentLevel != null && isNumber(options.indentLevel)
      ? options.indentLevel
      : 0
  let _content = ''

  function push(content: string): void {
    _content += content
  }

  function _newline(n: number) {
    push('\n' + `  `.repeat(n))
  }

  function newline() {
    _newline(_indentLevel)
  }

  function pushline(content: string) {
    push(content)
    newline()
  }

  function indent(withoutNewLine?: boolean) {
    if (withoutNewLine) {
      ++_indentLevel
    } else {
      _newline(++_indentLevel)
    }
  }

  function deindent(withoutNewLine?: boolean) {
    if (withoutNewLine) {
      --_indentLevel
    } else {
      _newline(--_indentLevel)
    }
  }

  return {
    get indentLevel(): number {
      return _indentLevel
    },
    get content(): string {
      return String(_content)
    },
    push,
    newline,
    pushline,
    indent,
    deindent
  }
}
