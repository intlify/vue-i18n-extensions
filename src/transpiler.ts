import { parse } from '@babel/parser'
import type { ObjectProperty, ObjectExpression, Expression } from '@babel/types'

interface EvaluateReturn {
  status: 'ok' | 'ng'
  value?: unknown
}

export type NestedValue<T> = { [P in keyof T]: NestedValue<T[P]> }

export type TranslationParams<T = Record<string, unknown>> = {
  path: string
  named: NestedValue<T>
  options: {
    locale?: string
    plural?: string
  }
}

export function evaluateValue(expression: string): EvaluateReturn {
  const ret = { status: 'ng', value: undefined } as EvaluateReturn

  try {
    const ast = parse(`const a = ${expression.trim()}`)
    if (
      ast.program.body.length >= 1 &&
      ast.program.body[0].type === 'VariableDeclaration' &&
      ast.program.body[0].declarations.length >= 1
    ) {
      const node = ast.program.body[0].declarations[0].init
      if (node != null && (node.type === 'StringLiteral' || node.type === 'ObjectExpression')) {
        // eslint-disable-next-line @typescript-eslint/no-implied-eval
        const val = new Function(`return ${expression.trim()} `)() as unknown
        ret.status = 'ok'
        ret.value = val
      }
    }
    // eslint-disable-next-line no-empty
  } catch (_e) {}

  return ret
}

export function parseVTExpression(expression: string): TranslationParams {
  const ret: TranslationParams = {
    path: '',
    named: {},
    options: {}
  }

  try {
    const ast = parse(`const a = ${expression.trim()}`)
    if (
      ast.program.body.length >= 1 &&
      ast.program.body[0].type === 'VariableDeclaration' &&
      ast.program.body[0].declarations.length >= 1
    ) {
      const node = ast.program.body[0].declarations[0].init
      if (node != null) {
        if (node.type === 'StringLiteral' && node.extra) {
          ret.path = node.extra.raw as string
        } else if (node.type === 'Identifier') {
          ret.path = node.name
        } else if (node.type === 'MemberExpression') {
          ret.path = getObjectMemberValue(node)
        } else if (node.type === 'ObjectExpression') {
          node.properties.forEach(propNode => {
            if (propNode.type === 'ObjectProperty') {
              const propKeyNode = propNode.key
              if (propKeyNode.type !== 'Identifier') {
                return
              }
              const propValueNode = propNode.value
              switch (propKeyNode.name) {
                case 'path':
                  ret.path = getObjectMemberValue(propValueNode)
                  break
                case 'locale':
                  ret.options.locale = getObjectMemberValue(propValueNode)
                  break
                case 'choice':
                case 'plural':
                  ret.options.plural = getObjectMemberValue(propValueNode)
                  break
                case 'args':
                  // console.log('args', propValueNode)
                  if (propValueNode.type === 'ObjectExpression') {
                    traverseObjectMember(propValueNode, ret.named)
                  }
                  break
                default:
                  break
              }
            }
          })
        }
      }
    }
    // eslint-disable-next-line no-empty
  } catch (_e) {}

  return ret
}

function getObjectMemberValue(node: ObjectProperty['value']): string {
  if ((node.type === 'StringLiteral' || node.type === 'NumericLiteral') && node.extra) {
    return node.extra.raw as string
  } else if (node.type === 'Identifier') {
    return node.name
  } else if (node.type === 'MemberExpression') {
    const paths: string[] = []
    collectMemberPath(node, paths)
    paths.reverse()
    return paths.join('.')
  } else {
    return ''
  }
}

function traverseObjectMember(
  node: ObjectExpression,
  target: NestedValue<Record<string, unknown>>
): void {
  node.properties.forEach(propNode => {
    if (propNode.type === 'ObjectProperty') {
      const propKeyNode = propNode.key
      if (propKeyNode.type !== 'Identifier') {
        return
      }
      if (!(propKeyNode.name in target)) {
        target[propKeyNode.name] = {}
      }
      // console.log('propNode', propNode)
      const propValueNode = propNode.value
      if (propValueNode.type === 'ObjectExpression') {
        traverseObjectMember(propValueNode, target[propKeyNode.name])
      } else {
        target[propKeyNode.name] = getObjectMemberValue(propValueNode)
      }
    }
  })
}

function collectMemberPath(node: Expression, paths: string[]): void {
  if (node.type === 'Identifier') {
    paths.push(node.name)
    return
  } else if (node.type === 'MemberExpression' && node.property.type === 'Identifier') {
    paths.push(node.property.name)
    return collectMemberPath(node.object, paths)
  }
}
