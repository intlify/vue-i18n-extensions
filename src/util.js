const stringRE = /'(?:[^'\\]|\\.)*'|"(?:[^"\\]|\\.)*"|`(?:[^`\\]|\\.)*\$\{|\}(?:[^`\\]|\\.)*`|`(?:[^`\\]|\\.)*`/g
const ecmaKeywordsRE = new RegExp('(?<!\\.)\\b' + (
  'delete,typeof,instanceof,void,do,if,for,let,new,try,var,case,else,with,await,break,catch,class,const,' +
  'alert,eval,super,throw,while,yield,delete,export,import,return,switch,default,' +
  'extends,finally,continue,debugger,function,arguments'
).split(',').join('\\b(?!\\.)|(?<!\\.)\\b') + '\\b(?!\\.)')

function warn (msg, err) {
  if (typeof console !== 'undefined') {
    console.warn('[vue-i18n-extensions] ' + msg)
    /* istanbul ignore if */
    if (err) {
      console.warn(err.stack)
    }
  }
}

const toString = Object.prototype.toString
const OBJECT_STRING = '[object Object]'
function isPlainObject (obj) {
  return toString.call(obj) === OBJECT_STRING
}

function addProp (el, name, value) {
  (el.props || (el.props = [])).push({ name, value })
}

function getAttr (el, name) {
  return el.attrsMap[name]
}

function removeAttr (el, name) {
  if ((el.attrsMap[name]) !== null) {
    const list = el.attrsList
    for (let i = 0, l = list.length; i < l; i++) {
      if (list[i].name === name) {
        list.splice(i, 1)
        break
      }
    }
  }
}

function evaluateValue (expression) {
  const ret = { status: 'ng', value: undefined }

  if (expression.match(ecmaKeywordsRE)) { return ret }
  if (!expression.match(stringRE)) { return ret }

  try {
    const val = (new Function(`return ${expression}`))()
    ret.status = 'ok'
    ret.value = val
  } catch (e) { }
  
  return ret
}

module.exports = {
  warn,
  isPlainObject,
  addProp,
  getAttr,
  removeAttr,
  evaluateValue
}
