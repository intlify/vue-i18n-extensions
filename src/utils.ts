import { parse } from '@babel/parser'

interface EvaluateReturn {
  status: 'ok' | 'ng'
  value?: unknown
}

export function evaluateValue(expression: string): EvaluateReturn {
  const ret = { status: 'ng', value: undefined } as EvaluateReturn

  try {
    const ast = parse(`const a = ${expression.trim()}`)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const node = (ast.program.body[0] as any).declarations[0].init
    if (node.type === 'StringLiteral' || node.type === 'ObjectExpression') {
      const val = new Function(`return ${expression.trim()}`)()
      ret.status = 'ok'
      ret.value = val
    }
  } catch (e) {}

  return ret
}
