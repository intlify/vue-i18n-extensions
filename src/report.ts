import { format } from './utils'
import { SourceLocation } from '@vue/compiler-core'
import { isString } from '@vue/shared'

export interface ExtensionsError extends SyntaxError {
  code: number
  loc?: SourceLocation
}

export interface ReportOptions {
  mode?: 'error' | 'warn'
  loc?: SourceLocation
  args?: unknown[]
}

export const enum ReportCodes {
  // Special value for higher-order compilers to pick up the last code
  // to avoid collision of error codes. This should always be kept as the last
  // item.
  SUCCESS,
  UNEXPECTED_DIRECTIVE_EXPRESSION,
  NOT_SUPPORTED_BINDING_PRE_TRANSLATION,
  FAILED_VALUE_EVALUATION,
  REQUIRED_PARAMETER,
  INVALID_PARAMETER_TYPE,
  NOT_SUPPORTED_PRESERVE,
  ORVERRIDE_ELEMENT_CHILDREN,
  __EXTEND_POINT__
}

// TODO: should be extracted as i18n resources (intlify project self hosting!)
const ReportMessages: { [code: number]: string } = {
  [ReportCodes.UNEXPECTED_DIRECTIVE_EXPRESSION]: `Unexpected directive expression: {0}`,
  [ReportCodes.NOT_SUPPORTED_BINDING_PRE_TRANSLATION]: `Not support binding in pre-localization: {0}`,
  [ReportCodes.FAILED_VALUE_EVALUATION]: `Failed valu evaluation: {0}`,
  [ReportCodes.REQUIRED_PARAMETER]: `Required parameter: {0}`,
  [ReportCodes.INVALID_PARAMETER_TYPE]: `Required parameter: {0}`,
  [ReportCodes.NOT_SUPPORTED_PRESERVE]: `Not supportted 'preserve': {0}`,
  [ReportCodes.ORVERRIDE_ELEMENT_CHILDREN]: `v-t will override element children: {0}`
}

export function getReportMessage(code: ReportCodes, ...args: unknown[]): string {
  return format(ReportMessages[code], ...args)
}

function createExtensionsError(
  code: ReportCodes,
  msg: string,
  loc?: SourceLocation
): ExtensionsError {
  const error = new SyntaxError(msg) as ExtensionsError
  error.code = code
  if (loc) {
    error.loc = loc
  }
  return error
}

export function report(
  code: ReportCodes,
  optinos: ReportOptions = {}
): void | ExtensionsError {
  const mode =
    optinos.mode &&
    isString(optinos.mode) &&
    ['warn', 'error'].includes(optinos.mode)
      ? optinos.mode
      : 'warn'

  const msg = getReportMessage(code, optinos.args)
  if (mode === 'warn') {
    console.warn('[vue-i18n-extensions] ' + msg)
  } else {
    // error
    throw createExtensionsError(code, msg, optinos.loc)
  }
}
