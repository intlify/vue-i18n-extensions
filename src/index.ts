import {
  DirectiveTransform,
  createObjectProperty,
  createSimpleExpression,
  SimpleExpressionNode
} from '@vue/compiler-dom'
import { isObject, isString, toDisplayString } from '@vue/shared'
import { I18n, Locale } from 'vue-i18n'
import { evaluateValue } from './transpiler'
import { report, ReportCodes } from './report'

// TODO: should be imported from vue-i18n
type VTDirectiveValue = {
  path: string
  locale?: Locale
  args?: { [prop: string]: unknown }
  choice?: number
}

// TODO: should be used from shared libs
export const isNumber = (val: unknown): val is number =>
  typeof val === 'number' && isFinite(val)

export function defineTransformVT(i18n: I18n): DirectiveTransform {
  return (dir, node /*, context*/) => {
    const { exp, loc } = dir
    if (!exp) {
      // TODO: throw error with context.OnError
      // NOTE: We need to support from @vue/compiler-core
      // https://github.com/vuejs/vue-next/issues/1147
      report(ReportCodes.UNEXPECTED_DIRECTIVE_EXPRESSION, {
        mode: 'error',
        args: [node.loc.source || ''],
        loc: node.loc
      })
    }
    if (node.children.length > 0) {
      // TODO: throw error with context.OnError
      // NOTE: We need to support from @vue/compiler-core
      // https://github.com/vuejs/vue-next/issues/1147
      report(ReportCodes.ORVERRIDE_ELEMENT_CHILDREN, {
        mode: 'error',
        args: [node.loc.source || ''],
        loc: node.loc
      })
      node.children.length = 0
    }

    const simpleExp = exp as SimpleExpressionNode
    if (simpleExp.isConstant) {
      if (dir.modifiers.includes('preserve')) {
        report(ReportCodes.NOT_SUPPORTED_PRESERVE, {
          args: [node.loc.source || ''],
          loc: node.loc
        })
      }

      const { status, value } = evaluateValue(simpleExp.content)
      if (status === 'ng') {
        report(ReportCodes.FAILED_VALUE_EVALUATION, {
          args: [node.loc.source || ''],
          loc: node.loc
        })
        return { props: [] }
      }

      const [parsedValue, parseStatus] = parseValue(value)
      if (parseStatus !== ReportCodes.SUCCESS) {
        report(parseStatus, { args: [node.loc.source || ''], loc: node.loc })
        return { props: [] }
      }

      const translated = i18n.global.t(...makeParams(parsedValue!))
      if (translated) {
        simpleExp.content = toDisplayString(`${JSON.stringify(translated)}`)
      }

      // success!
      return {
        props: [
          createObjectProperty(
            createSimpleExpression(`textContent`, true, loc),
            exp || createSimpleExpression('', true)
          )
        ]
      }
    } else {
      report(ReportCodes.NOT_SUPPORTED_BINDING_PRE_TRANSLATION, {
        args: [node.loc.source || ''],
        loc: node.loc
      })
      return { props: [] }
    }
  }
}

// TODO: should be defined typing ...
function parseValue(value: unknown): [VTDirectiveValue | null, ReportCodes] {
  if (isString(value)) {
    return [{ path: value }, ReportCodes.SUCCESS]
  } else if (isObject(value)) {
    if (!('path' in value)) {
      return [null, ReportCodes.REQUIRED_PARAMETER]
    }
    return [value as VTDirectiveValue, ReportCodes.SUCCESS]
  } else {
    return [null, ReportCodes.INVALID_PARAMETER_TYPE]
  }
}

function makeParams(value: VTDirectiveValue): unknown[] {
  const { path, locale, args, choice } = value
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const options = {} as any
  const named: { [prop: string]: unknown } = args || {}

  if (isString(locale)) {
    options.locale = locale
  }

  if (isNumber(choice)) {
    options.plural = choice
  }

  return [path, named, options]
}
