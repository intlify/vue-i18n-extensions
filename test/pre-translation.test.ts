import { compile } from '@vue/compiler-dom'
import { createI18n } from 'vue-i18n'
import { transformVTDirective } from '../src/transform'

let spyWarn: any // eslint-disable-line @typescript-eslint/no-explicit-any
beforeEach(() => {
  spyWarn = jest.spyOn(global.console, 'warn')
})

afterEach(() => {
  // jest.clearAllMocks()
  // jest.resetAllMocks()
  jest.restoreAllMocks()
})

test('basic', () => {
  const i18n = createI18n({
    locale: 'ja',
    messages: {
      ja: { hello: 'こんにちは！' },
      en: { hello: 'hello!' }
    }
  })
  const transformVT = transformVTDirective({ i18n })
  const source = `<div v-t="'hello'"/>`
  const { code, ast } = compile(source, {
    mode: 'function',
    hoistStatic: true,
    prefixIdentifiers: true,
    directiveTransforms: { t: transformVT }
  })
  expect(code).toMatchSnapshot(source)
  expect(ast).toMatchSnapshot(source)
})

test('named', () => {
  const i18n = createI18n({
    locale: 'en',
    messages: {
      en: { hello: 'hello, {name}!' },
      ja: { hello: 'こんにちは、{name}！' }
    }
  })
  const transformVT = transformVTDirective({ i18n })
  const source = `<div v-t="{ path: 'hello', locale: 'ja', args: { name: 'kazupon' } }"/>`
  const { code, ast } = compile(source, {
    mode: 'function',
    hoistStatic: true,
    prefixIdentifiers: true,
    directiveTransforms: { t: transformVT }
  })
  expect(code).toMatchSnapshot(source)
  expect(ast).toMatchSnapshot(source)
})

test('plural', () => {
  const i18n = createI18n({
    locale: 'en',
    messages: {
      en: {
        banana: 'no bananas | {n} banana | {n} bananas'
      }
    }
  })

  const transformVT = transformVTDirective({ i18n })
  const source = `<div v-t="{ path: 'banana', choice: 2 }"/>`
  const { code, ast } = compile(source, {
    mode: 'function',
    hoistStatic: true,
    prefixIdentifiers: true,
    directiveTransforms: { t: transformVT }
  })
  expect(code).toMatchSnapshot(source)
  expect(ast).toMatchSnapshot(source)
})

test('missing translation', () => {
  spyWarn.mockImplementation(() => {}) // eslint-disable-line @typescript-eslint/no-empty-function

  const i18n = createI18n({
    legacy: false,
    locale: 'ja',
    messages: {
      ja: {}
    }
  })
  const transformVT = transformVTDirective({ i18n })
  const source = `<div v-t="'foo.bar'"/>`
  const { code, ast } = compile(source, {
    mode: 'function',
    hoistStatic: true,
    prefixIdentifiers: true,
    directiveTransforms: { t: transformVT }
  })
  expect(code).toMatchSnapshot(source)
  expect(ast).toMatchSnapshot(source)
  expect(spyWarn.mock.calls[0][0]).toEqual(
    `[intlify] Not found 'foo.bar' key in 'ja' locale messages.`
  )
})
