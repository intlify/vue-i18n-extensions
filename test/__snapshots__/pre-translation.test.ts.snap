// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic: <div v-t="'hello'"/> 1`] = `
"const { openBlock: _openBlock, createBlock: _createBlock } = Vue

return function render(_ctx, _cache) {
  return (_openBlock(), _createBlock(\\"div\\", null, \\"こんにちは！\\"))
}"
`;

exports[`basic: <div v-t="'hello'"/> 2`] = `
Object {
  "cached": 0,
  "children": Array [
    Object {
      "children": Array [
        Object {
          "content": "こんにちは！",
          "type": 2,
        },
      ],
      "codegenNode": Object {
        "children": Object {
          "content": "こんにちは！",
          "type": 2,
        },
        "directives": undefined,
        "disableTracking": false,
        "dynamicProps": undefined,
        "isBlock": true,
        "loc": Object {
          "end": Object {
            "column": 21,
            "line": 1,
            "offset": 20,
          },
          "source": "<div v-t=\\"'hello'\\"/>",
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "patchFlag": undefined,
        "props": undefined,
        "tag": "\\"div\\"",
        "type": 13,
      },
      "isSelfClosing": true,
      "loc": Object {
        "end": Object {
          "column": 21,
          "line": 1,
          "offset": 20,
        },
        "source": "<div v-t=\\"'hello'\\"/>",
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "ns": 0,
      "props": Array [
        Object {
          "arg": undefined,
          "exp": Object {
            "constType": 3,
            "content": "'hello'",
            "identifiers": Array [],
            "isStatic": false,
            "loc": Object {
              "end": Object {
                "column": 18,
                "line": 1,
                "offset": 17,
              },
              "source": "'hello'",
              "start": Object {
                "column": 11,
                "line": 1,
                "offset": 10,
              },
            },
            "type": 4,
          },
          "loc": Object {
            "end": Object {
              "column": 19,
              "line": 1,
              "offset": 18,
            },
            "source": "v-t=\\"'hello'\\"",
            "start": Object {
              "column": 6,
              "line": 1,
              "offset": 5,
            },
          },
          "modifiers": Array [],
          "name": "t",
          "type": 7,
        },
      ],
      "tag": "div",
      "tagType": 0,
      "type": 1,
    },
  ],
  "codegenNode": Object {
    "children": Object {
      "content": "こんにちは！",
      "type": 2,
    },
    "directives": undefined,
    "disableTracking": false,
    "dynamicProps": undefined,
    "isBlock": true,
    "loc": Object {
      "end": Object {
        "column": 21,
        "line": 1,
        "offset": 20,
      },
      "source": "<div v-t=\\"'hello'\\"/>",
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "patchFlag": undefined,
    "props": undefined,
    "tag": "\\"div\\"",
    "type": 13,
  },
  "components": Array [],
  "directives": Array [],
  "helpers": Array [
    Symbol(openBlock),
    Symbol(createBlock),
  ],
  "hoists": Array [],
  "imports": Array [],
  "loc": Object {
    "end": Object {
      "column": 21,
      "line": 1,
      "offset": 20,
    },
    "source": "<div v-t=\\"'hello'\\"/>",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "temps": 0,
  "type": 0,
}
`;

exports[`missing translation: <div v-t="'foo.bar'"/> 1`] = `
"const { openBlock: _openBlock, createBlock: _createBlock } = Vue

return function render(_ctx, _cache) {
  return (_openBlock(), _createBlock(\\"div\\", null, \\"foo.bar\\"))
}"
`;

exports[`missing translation: <div v-t="'foo.bar'"/> 2`] = `
Object {
  "cached": 0,
  "children": Array [
    Object {
      "children": Array [
        Object {
          "content": "foo.bar",
          "type": 2,
        },
      ],
      "codegenNode": Object {
        "children": Object {
          "content": "foo.bar",
          "type": 2,
        },
        "directives": undefined,
        "disableTracking": false,
        "dynamicProps": undefined,
        "isBlock": true,
        "loc": Object {
          "end": Object {
            "column": 23,
            "line": 1,
            "offset": 22,
          },
          "source": "<div v-t=\\"'foo.bar'\\"/>",
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "patchFlag": undefined,
        "props": undefined,
        "tag": "\\"div\\"",
        "type": 13,
      },
      "isSelfClosing": true,
      "loc": Object {
        "end": Object {
          "column": 23,
          "line": 1,
          "offset": 22,
        },
        "source": "<div v-t=\\"'foo.bar'\\"/>",
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "ns": 0,
      "props": Array [
        Object {
          "arg": undefined,
          "exp": Object {
            "constType": 0,
            "content": "'foo.bar'",
            "identifiers": Array [],
            "isStatic": false,
            "loc": Object {
              "end": Object {
                "column": 20,
                "line": 1,
                "offset": 19,
              },
              "source": "'foo.bar'",
              "start": Object {
                "column": 11,
                "line": 1,
                "offset": 10,
              },
            },
            "type": 4,
          },
          "loc": Object {
            "end": Object {
              "column": 21,
              "line": 1,
              "offset": 20,
            },
            "source": "v-t=\\"'foo.bar'\\"",
            "start": Object {
              "column": 6,
              "line": 1,
              "offset": 5,
            },
          },
          "modifiers": Array [],
          "name": "t",
          "type": 7,
        },
      ],
      "tag": "div",
      "tagType": 0,
      "type": 1,
    },
  ],
  "codegenNode": Object {
    "children": Object {
      "content": "foo.bar",
      "type": 2,
    },
    "directives": undefined,
    "disableTracking": false,
    "dynamicProps": undefined,
    "isBlock": true,
    "loc": Object {
      "end": Object {
        "column": 23,
        "line": 1,
        "offset": 22,
      },
      "source": "<div v-t=\\"'foo.bar'\\"/>",
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "patchFlag": undefined,
    "props": undefined,
    "tag": "\\"div\\"",
    "type": 13,
  },
  "components": Array [],
  "directives": Array [],
  "helpers": Array [
    Symbol(openBlock),
    Symbol(createBlock),
  ],
  "hoists": Array [],
  "imports": Array [],
  "loc": Object {
    "end": Object {
      "column": 23,
      "line": 1,
      "offset": 22,
    },
    "source": "<div v-t=\\"'foo.bar'\\"/>",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "temps": 0,
  "type": 0,
}
`;

exports[`named: <div v-t="{ path: 'hello', locale: 'ja', args: { name: 'kazupon' } }"/> 1`] = `
"const { openBlock: _openBlock, createBlock: _createBlock } = Vue

return function render(_ctx, _cache) {
  return (_openBlock(), _createBlock(\\"div\\", null, \\"こんにちは、kazupon！\\"))
}"
`;

exports[`named: <div v-t="{ path: 'hello', locale: 'ja', args: { name: 'kazupon' } }"/> 2`] = `
Object {
  "cached": 0,
  "children": Array [
    Object {
      "children": Array [
        Object {
          "content": "こんにちは、kazupon！",
          "type": 2,
        },
      ],
      "codegenNode": Object {
        "children": Object {
          "content": "こんにちは、kazupon！",
          "type": 2,
        },
        "directives": undefined,
        "disableTracking": false,
        "dynamicProps": undefined,
        "isBlock": true,
        "loc": Object {
          "end": Object {
            "column": 72,
            "line": 1,
            "offset": 71,
          },
          "source": "<div v-t=\\"{ path: 'hello', locale: 'ja', args: { name: 'kazupon' } }\\"/>",
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "patchFlag": undefined,
        "props": undefined,
        "tag": "\\"div\\"",
        "type": 13,
      },
      "isSelfClosing": true,
      "loc": Object {
        "end": Object {
          "column": 72,
          "line": 1,
          "offset": 71,
        },
        "source": "<div v-t=\\"{ path: 'hello', locale: 'ja', args: { name: 'kazupon' } }\\"/>",
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "ns": 0,
      "props": Array [
        Object {
          "arg": undefined,
          "exp": Object {
            "constType": 3,
            "content": "{ path: 'hello', locale: 'ja', args: { name: 'kazupon' } }",
            "identifiers": Array [],
            "isStatic": false,
            "loc": Object {
              "end": Object {
                "column": 69,
                "line": 1,
                "offset": 68,
              },
              "source": "{ path: 'hello', locale: 'ja', args: { name: 'kazupon' } }",
              "start": Object {
                "column": 11,
                "line": 1,
                "offset": 10,
              },
            },
            "type": 4,
          },
          "loc": Object {
            "end": Object {
              "column": 70,
              "line": 1,
              "offset": 69,
            },
            "source": "v-t=\\"{ path: 'hello', locale: 'ja', args: { name: 'kazupon' } }\\"",
            "start": Object {
              "column": 6,
              "line": 1,
              "offset": 5,
            },
          },
          "modifiers": Array [],
          "name": "t",
          "type": 7,
        },
      ],
      "tag": "div",
      "tagType": 0,
      "type": 1,
    },
  ],
  "codegenNode": Object {
    "children": Object {
      "content": "こんにちは、kazupon！",
      "type": 2,
    },
    "directives": undefined,
    "disableTracking": false,
    "dynamicProps": undefined,
    "isBlock": true,
    "loc": Object {
      "end": Object {
        "column": 72,
        "line": 1,
        "offset": 71,
      },
      "source": "<div v-t=\\"{ path: 'hello', locale: 'ja', args: { name: 'kazupon' } }\\"/>",
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "patchFlag": undefined,
    "props": undefined,
    "tag": "\\"div\\"",
    "type": 13,
  },
  "components": Array [],
  "directives": Array [],
  "helpers": Array [
    Symbol(openBlock),
    Symbol(createBlock),
  ],
  "hoists": Array [],
  "imports": Array [],
  "loc": Object {
    "end": Object {
      "column": 72,
      "line": 1,
      "offset": 71,
    },
    "source": "<div v-t=\\"{ path: 'hello', locale: 'ja', args: { name: 'kazupon' } }\\"/>",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "temps": 0,
  "type": 0,
}
`;

exports[`plural: <div v-t="{ path: 'banana', choice: 2 }"/> 1`] = `
"const { openBlock: _openBlock, createBlock: _createBlock } = Vue

return function render(_ctx, _cache) {
  return (_openBlock(), _createBlock(\\"div\\", null, \\"2 bananas\\"))
}"
`;

exports[`plural: <div v-t="{ path: 'banana', choice: 2 }"/> 2`] = `
Object {
  "cached": 0,
  "children": Array [
    Object {
      "children": Array [
        Object {
          "content": "2 bananas",
          "type": 2,
        },
      ],
      "codegenNode": Object {
        "children": Object {
          "content": "2 bananas",
          "type": 2,
        },
        "directives": undefined,
        "disableTracking": false,
        "dynamicProps": undefined,
        "isBlock": true,
        "loc": Object {
          "end": Object {
            "column": 43,
            "line": 1,
            "offset": 42,
          },
          "source": "<div v-t=\\"{ path: 'banana', choice: 2 }\\"/>",
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "patchFlag": undefined,
        "props": undefined,
        "tag": "\\"div\\"",
        "type": 13,
      },
      "isSelfClosing": true,
      "loc": Object {
        "end": Object {
          "column": 43,
          "line": 1,
          "offset": 42,
        },
        "source": "<div v-t=\\"{ path: 'banana', choice: 2 }\\"/>",
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "ns": 0,
      "props": Array [
        Object {
          "arg": undefined,
          "exp": Object {
            "constType": 3,
            "content": "{ path: 'banana', choice: 2 }",
            "identifiers": Array [],
            "isStatic": false,
            "loc": Object {
              "end": Object {
                "column": 40,
                "line": 1,
                "offset": 39,
              },
              "source": "{ path: 'banana', choice: 2 }",
              "start": Object {
                "column": 11,
                "line": 1,
                "offset": 10,
              },
            },
            "type": 4,
          },
          "loc": Object {
            "end": Object {
              "column": 41,
              "line": 1,
              "offset": 40,
            },
            "source": "v-t=\\"{ path: 'banana', choice: 2 }\\"",
            "start": Object {
              "column": 6,
              "line": 1,
              "offset": 5,
            },
          },
          "modifiers": Array [],
          "name": "t",
          "type": 7,
        },
      ],
      "tag": "div",
      "tagType": 0,
      "type": 1,
    },
  ],
  "codegenNode": Object {
    "children": Object {
      "content": "2 bananas",
      "type": 2,
    },
    "directives": undefined,
    "disableTracking": false,
    "dynamicProps": undefined,
    "isBlock": true,
    "loc": Object {
      "end": Object {
        "column": 43,
        "line": 1,
        "offset": 42,
      },
      "source": "<div v-t=\\"{ path: 'banana', choice: 2 }\\"/>",
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "patchFlag": undefined,
    "props": undefined,
    "tag": "\\"div\\"",
    "type": 13,
  },
  "components": Array [],
  "directives": Array [],
  "helpers": Array [
    Symbol(openBlock),
    Symbol(createBlock),
  ],
  "hoists": Array [],
  "imports": Array [],
  "loc": Object {
    "end": Object {
      "column": 43,
      "line": 1,
      "offset": 42,
    },
    "source": "<div v-t=\\"{ path: 'banana', choice: 2 }\\"/>",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "temps": 0,
  "type": 0,
}
`;
